Space invaders blockly - first draft

jessica-norris opened the pull request:
First issue of blocks for space invaders course consisting of the following:

- block to create an img element using a specific drop-down set of images
- block to set the position [top/left] of an element (using the style) to a certain number of pixels
- block to change the position [top/left] of an element by a certain number of pixels (may want to modify to look like the change-variable block)

As a side-effect of this, the `getAllSpaceships` block is using code-gen for a more generic `querySelectorAll` so I've added this as a block just like the existing `querySelector` block -- this returns a list.

I've asked for review from @taybenlor (to get an opinion about the change position block) and @jimmo, to spread the review load wherever someone has time. I have more blocks to do that relate to calling custom functions that will release on this same branch.

See example:

<img width="408" alt="screen shot 2018-04-30 at 11 17 48 am" src="https://user-images.githubusercontent.com/7776025/39413047-33628172-4c68-11e8-9fbd-d666212e0a36.png">

With code-gen looking like:

<img width="529" alt="screen shot 2018-04-30 at 11 17 59 am" src="https://user-images.githubusercontent.com/7776025/39413050-3891e39a-4c68-11e8-94c8-76f5b04bfe04.png">

Fixes groklearning/issues#341

jimmo commented on blocks/dom-elements.js:
```diff
@@ -444,4 +464,203 @@ Blockly.defineBlocksWithJsonArray([
   },
+  // Custom space invaders block like the <script> block.
+  {
+    "type": "space_invaders_script_element",
```
Maybe rename to "space_invaders_in_space" or "space_invaders_in_space_script_element" just to be clear that it's the "in space" block. (And not like a special custom script tag for space invaders)

jimmo commented on generators/html/dom-elements.js:
```diff
@@ -103,4 +103,27 @@ Blockly.HTML['cookie_clicker_src_url_picker'] = function(block) {
+  var js = Blockly.JavaScript.statementToCode(block, 'CONTENT') || '';
+  // Note: strip leading newlines because the JavaScript codegen will always join the definitions to the
+  // code with \n\n\n even if there are no definitions.
+  js = Blockly.JavaScript.finish(js).replace(/^\n*/,'');
```
I think this would be clearer as `js = Blockly.JavaScript.finish(js).trim()`

jimmo replied:
Also it's probably worth overriding Blockly.JavaScript.finish to replace `\n+` with `\n` anywhere in the string.

jessica-norris replied:
Done, I think.

jimmo commented on generators/javascript/dom-elements.js:
```diff
@@ -40,4 +47,47 @@ Blockly.JavaScript['cookie_clicker_img_url_picker'] = function(block) {
 
+/** Create a new DOM <img> element with the given src url. */
+Blockly.JavaScript['space_invaders_new_ship'] = function(block) {
+  var url = block.getFieldValue('URL');
```
Can you use `valueToCode` instead of `getFieldValue` to avoid the need to call `quote_` directly?

jessica-norris replied:
I don't think I can here, because the field being extracted is a drop-down of values, so there is no code-gen per se. The `getFieldValue` is getting literals out of the drop-down.

I could simple ensure that the drop-downs are quoted, though, I suppose? (In the block definition).

jimmo commented on generators/javascript/loops.js:
```diff
@@ -108,7 +108,7 @@ Blockly.JavaScript['loops_read_ask'] = function(block) {
 
 // Override default generator for forEach to use ES6 for/of loop.
-Blockly.JavaScript['controls_forEach'] = function(block) {
+Blockly.JavaScript['loops_forEach'] = function(block) {
```
This is a bit confusing -- the blocks names are `controls_foo`, the `loops_` are aliases. So to match the block definition, all the generators should be `Blockly.JavaScript['controls_foo']`, then the aliasing `ORIGINAL_LOOP_TYPES` from the top of the file should be moved to the end to also include the ones defined in this file.

The effect of the current behaviour (that this will fix) is that when we override an existing one, the loops_foo and controls_foo are no longer the same.

jessica-norris replied:
This is interesting -- I was fixing a bug here (or I thought I was). I thought that current (using `controls_` this was not correctly picking up and overriding the code gen because they get renamed (aliased) before we have a chance to override them. I guess what I should do is:

Override the `controls_forEach` and then re-alias the copy, so that the alias is also changed?

jimmo replied:
Yes - the entire file should be setting `Blockly.JavaScript['controls_foo']` for all blocks, and then the `ORIGINAL_LOOP_TYPES` should be updated to be more like the one at the end of blocks/loops.js and at the end of the file.

jessica-norris replied:
This is now done.

jessica-norris commented:
I think this is now looking pretty solid. Karsten also approved the `requestAnimationFrame` technique, so we can include the new blocks as follows:

* 'on next animation frame'
* 'starting animating with'

<img width="426" alt="screen shot 2018-04-30 at 2 19 40 pm" src="https://user-images.githubusercontent.com/7776025/39459091-629870da-4d3c-11e8-84ff-c534c3adc72c.png">

jimmo commented on generators/javascript.js:
```diff
@@ -6,5 +6,11 @@ goog.require('Blockly.JavaScript');
 
 Blockly.JavaScript.addReservedWords('await,of,arguments');
 
+// Override finish() to strip leading whitespace that result from an empty definitions section.
```
The convention has been to use `name_` and generally it's best in JS to use `apply` so that `this` is correct. (In this case it doesn't matter, but if you always use `apply` it can't go wrong). See https://github.com/groklearning/blockly-blocks/blob/develop/generators/python.js for an example.

jessica-norris replied:
Fixed

jimmo commented on blocks/space-invaders.js:
```diff
@@ -0,0 +1,243 @@
+              "height": 15,
+              "alt": "*"
+            },
+            "spaceship2.png"
```
Is there a spaceship1.png?

jessica-norris replied:
Nice spotting. The names don't align with the numbering on the CDN, but I've verified these are the correct images and now they are down with data urls. So all good. :)

jessica-norris commented:
I think this has now addressed remaining review issues. Are we all good with this? My one remaining question is whether Karsten will realise where the function is for defining `animate()` -- this is located in the "Events" tray and looks like an event handler.

@jimmo do you think it might be sufficient to get Karsten to supply this is a read-only block on the workspace, predefined? I can export the playground question I'm using to test the blocks here.
